lib/spell-check-manager.coffee

```diff
@@ -0,0 +1,306 @@
+    console.log "spell-check: addPluginChecker:", checker
+    @addSpellChecker checker
+  addSpellChecker: (checker) ->
+    console.log "spell-check: addSpellChecker:", checker
+    @checkers.push checker
+
+  removeSpellChecker: (spellChecker) ->
+    @checkers = @checkers.filter (plugin) -> plugin isnt spellChecker
+
+  check: (args, text) ->
+    # Make sure our deferred initialization is done.
+    @init()
+
+    # We need a couple packages.
+    multirange = require 'multi-integer-range'
+
+    # For every registered spellchecker, we need to find out the ranges in the
+    # text that the checker confirms are correct or indicates is a misspelling.
+    # We keep these as separate lists since the different checkers may indicate
+    # the same range for either and we need to be able to remove confirmed words
+    # from the misspelled ones.
+    correct = new multirange.MultiRange([])
+    incorrects = []
+
+    for checker in @checkers
+      # We only care if this plugin contributes to checking spelling.
+      if not checker.isEnabled() or not checker.providesSpelling(args)
+        continue
+
+      # Get the results which includes positive (correct) and negative (incorrect)
+      # ranges.
+      results = checker.check(args, text)
+
+      if results.correct
+        for range in results.correct
+          correct.appendRange(range.start, range.end)
+
+      if results.incorrect
+        newIncorrect = new multirange.MultiRange([])
+        incorrects.push(newIncorrect)
+
+        for range in results.incorrect
+          newIncorrect.appendRange(range.start, range.end)
+
+    # If we don't have any incorrect spellings, then there is nothing to worry
+    # about, so just return and stop processing.
+    misspellings = []
+
+    if incorrects.length is 0
+      return {id: args.id, misspellings}
+
+    # Build up an intersection of all the incorrect ranges. We only treat a word
+    # as being incorrect if *every* checker that provides negative values treats
+    # it as incorrect. We know there are at least one item in this list, so pull
+    # that out. If that is the only one, we don't have to do any additional work,
+    # otherwise we compare every other one against it, removing any elements
+    # that aren't an intersection which (hopefully) will produce a smaller list
+    # with each iteration.
+    intersection = null
+    index = 1
+
+    for incorrect in incorrects
+      if intersection is null
+        intersection = incorrect
+      else
+        intersection.intersect(incorrects[index])
+
+    # If we have no intersection, then nothing to report as a problem.
+    if intersection.length is 0
+      return {id: args.id, misspellings}
+
+    # Remove all of the confirmed correct words from the resulting incorrect
+    # list. This allows us to have correct-only providers as opposed to only
+    # incorrect providers.
+    if correct.ranges.length > 0
+      intersection.subtract(correct)
```

Comment:
Sorry, I think I should have clarified the meaning of the `Range` syntax I used. `Range` in Atom is defined as `Range(start: Point, end: Point)`: writing `Range([0, 1], [0, 3])` means `a range that spans from {row: 0, column: 1} to {row: 0, column: 3}`.

> From your example, I'm a bit confused on the input since I would expect the first to be [0, 20] and the second to be [0, 18] since there should never be an overlap in the resulting ranges.

If the contract is that two different checkers could never return overlapping ranges, and assuming that performance-wise crossing the native boundary could be fine, what do you think about enforcing it by using that word-based API I was talking about in the comment above?